### PR TITLE
Add parsed xml metadata to package info for find/install pkg

### PR DESCRIPTION
**Pull request by alerickson**
Parsed string xml in "nuspec" property of json file retrieved from VSTS feed.  
Added xml tags and values to package object so that package can successfully be found and installed.
___
Added condition (pkg.Published == null) in NugetLightRequest.cs to account for packages that do not have a publish date.

### CODE REVIEW

**Comment by alerickson** [comment]
I wanted to request a review of this PR as well but am unable to add reviewers here @bmanikm @jianyunt

File: Utility/DynamicJsonParser.cs [hunk 1]
```diff
@@ -135,6 +137,55 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
                     propertyName = propertyName.ToLowerInvariant();
                 }
 
+                // Nuspec contains xml metadata as a string (only applicable to VSTS feeds)
+                if (propertyName == "nuspec")
+                {
+                    XmlDocument xmlDoc = new XmlDocument();
+                    string xmlString = (string)psProperty.Value;
+
```
Comment (edyoung) [inline Utility/DynamicJsonParser.cs]: so the nuspec property contains xml but sometimes there is something else before the XML that we need to skip, is that right? That seems slightly odd - the xml parser should be able to cope with leading spaces, for example

File: Utility/DynamicJsonParser.cs [hunk 2]
```diff
@@ -135,6 +137,55 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                        var strIndex = xmlString.IndexOf('<');
+                        xmlString = xmlString.Substring(strIndex, xmlString.Length - 1);
+                    }
+
+                    xmlDoc.LoadXml(xmlString);
+                    XmlNodeList metadataNodeList = xmlDoc.GetElementsByTagName("metadata");
+                    List<string> propertyList = new List<string>();
+
+                    // Create a list of the child tags of metadata (id, version, etc.)
+                    foreach (XmlNode node in metadataNodeList)
```
Comment (edyoung) [inline Utility/DynamicJsonParser.cs]: generally we shouldn't need to parse the XML with regexes, it should be possible to get what we want direct from the xmlDoc object model. Could you include in a comment an example of the expected format of this nuspec tag? That would help explain what the code is doing

File: Utility/DynamicJsonParser.cs (hunk 1 above)
Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: Yes, I can't figure out why the Kusto Tools json file always adds a "?" before the  XML tag.  I haven't been able to reproduce this with other package jsons.

File: Utility/DynamicJsonParser.cs (hunk 2 above)
Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: Here's the contents of nuspec:
"nuspec":"﻿<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n<package xmlns=\"http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd\">\r\n  <metadata>\r\n    <id>Microsoft.Kusto.Tools</id>\r\n    <version>1.0.6</version>\r\n    <title>Kusto Tools</title>\r\n    <authors>Microsoft</authors>\r\n    <owners>Microsoft</owners>\r\n    <requireLicenseAcceptance>false</requireLicenseAcceptance>\r\n    <licenseUrl>http://kusto.blob.core.windows.net/kusto-nuget/EULA-agreement.htm</licenseUrl>\r\n    <projectUrl>http://aka.ms/kdocs</projectUrl>\r\n    <iconUrl>http://go.microsoft.com/fwlink/?LinkID=288890</iconUrl>\r\n    <description>Kusto Tools</description>\r\n    <releaseNotes>1.0.6: LightIngest refactored</releaseNotes>\r\n    <copyright>Copyright ©  Microsoft Corporation</copyright>\r\n  </metadata>\r\n</package>"

I initially had some issues pulling out the information I needed via xmlDoc, but I just tried again and was successfully able to get everything I need.  I'll push those changes.

File: Utility/DynamicJsonParser.cs [hunk 3]
```diff
@@ -135,6 +137,38 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
                     propertyName = propertyName.ToLowerInvariant();
                 }
 
+                // Nuspec contains xml metadata as a string (only applicable to VSTS feeds)
+                if (propertyName == "nuspec")
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: I think we have EqualsIgnoreCase extension. You may consider using it. e.g, propertyName.EqualsIgnoreCase("abc")

**Comment by jianyunt** [comment]
>         private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bool lowercaseProperties = true)

Could you please add comments for this GetDynamic method, including sample outputs like meta data etc. It's hard to parse in mind without sample?

---
Refers to: Utility/DynamicJsonParser.cs:106 in 3e0743e. [](commit_id = 3e0743e1a5fe61930aec70a7d9b454befc1f7a91, deletion_comment = False)

File: Utility/DynamicJsonParser.cs [hunk 4]
```diff
@@ -135,6 +137,38 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                {
+                    XmlDocument xmlDoc = new XmlDocument();
+                    string xmlString = (string)psProperty.Value;
+
+                    if (xmlString[0] != '<')
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: the meta data content is not well-formated xml?

File: Utility/DynamicJsonParser.cs [hunk 5]
```diff
@@ -21,6 +21,8 @@ namespace Microsoft.PackageManagement.NuGetProvider
     using System.Linq;
     using System.Management.Automation;
     using System.Reflection;
+    using System.Text.RegularExpressions;
+    using System.Xml;
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: Could you please add comments for this GetDynamic() method, including sample outputs like meta data etc. It's hard to parse in mind without sample?

File: Utility/DynamicJsonParser.cs (hunk 1 above)
Comment (edyoung) [reply Utility/DynamicJsonParser.cs]: Probably not really a ? but a non-printing character which is shown as ?.

File: Utility/DynamicJsonParser.cs (hunk 2 above)
Comment (edyoung) [reply Utility/DynamicJsonParser.cs]: Thanks - I meant a comment in the actual code so that people looking at the code in future can see it

**Review by edyoung — APPROVED** [review]

File: Utility/DynamicJsonParser.cs [hunk 6]
```diff
@@ -135,6 +137,38 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
                     propertyName = propertyName.ToLowerInvariant();
                 }
 
+                // Nuspec contains xml metadata as a string (only applicable to VSTS feeds)
+                if (propertyName == "nuspec")
+                {
+                    XmlDocument xmlDoc = new XmlDocument();
+                    string xmlString = (string)psProperty.Value;
```
Comment (bmanikm) [inline Utility/DynamicJsonParser.cs]: >xmlString [](start = 27, length = 9)

Consider checking for empty or whitespace. #Closed

File: Utility/DynamicJsonParser.cs [hunk 7]
```diff
@@ -135,6 +137,38 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                        if (node.HasChildNodes)
+                        {
+                            // Adding VSTS feed metadata values to the object containing nupkg property information
+                            for (int i = 0; i < node.ChildNodes.Count; i++)
+                            {
+                                string property = node.ChildNodes[i].Name;
+                                string propertyVal = node.ChildNodes[i].InnerText;
+
+                                object actualPropertyVal = ConvertObject(propertyVal);
+                                ((IDictionary<string, object>)actualObj)[property] = actualPropertyVal;
```
Comment (bmanikm) [inline Utility/DynamicJsonParser.cs]: >((IDictionary<string, object>)actualObj)[property] = actualPropertyVal; [](start = 32, length = 71)

Is this going to overwrite the earlier assigned property values in this loop? Consider assigning these values conditionally when the specified property is not available in the actualObj. #Closed

Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: Those values won't be overwriting anything because the json format for the VSTS feed only has those particular properties in nuspec.  The only properties for the json file are "commitId",  "previousCommitId", "nextCommitId", "createdDate", "modifiedDate", "nuspec",  and "_links".

I can add the condition in incase the json format changes in the future though. #Closed

File: Utility/DynamicJsonParser.cs [hunk 8]
```diff
@@ -135,6 +136,47 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                        throw new ArgumentNullException("xmlString");
+                    }
+
+                    if (xmlString[0] != '<')
+                    {
+                        var strIndex = xmlString.IndexOf('<');
+                        xmlString = xmlString.Substring(strIndex, xmlString.Length - 1);
+                    }
+
+                    xmlDoc.LoadXml(xmlString);
```
Comment (bmanikm) [inline Utility/DynamicJsonParser.cs]: >xmlDoc.LoadXml(xmlString); [](start = 20, length = 26)

Consider handling XmlException
[https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmldocument.loadxml?view=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmldocument.loadxml?view=netframework-4.7.2)

**Review by bmanikm — APPROVED** [review]
:shipit:

File: NugetLightRequest.cs [hunk 9]
```diff
@@ -1565,7 +1565,7 @@ private string ConfigurationFileLocation
                 // unlisted version is the one that has published year as 1900
                 if (!exactVersionRequired)
                 {
-                    pkgs = pkgs.Where(pkg => pkg.Published.HasValue && pkg.Published.Value.Year > 1900);
+                    pkgs = pkgs.Where(pkg => (!pkg.Published.HasValue || pkg.Published.Value.Year > 1900));
```
Comment (jianyunt) [inline NugetLightRequest.cs]: Can pkg.Published to be null?

**Review by jianyunt — CHANGES_REQUESTED** [review]

File: Utility/DynamicJsonParser.cs [hunk 10]
```diff
@@ -100,7 +101,7 @@ public static string Serialize(dynamic obj, string numericalPrefix = "n")
 
             return result.ImmediateBaseObject as string;
         }
-        
+
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs:105]: Could you please add comments for this GetDynamic() method, including sample outputs like meta data etc. It's hard to parse in mind without sample?

File: Utility/DynamicJsonParser.cs [hunk 11]
```diff
@@ -135,6 +136,47 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
 
+                // Nuspec contains xml metadata as a string (only applicable to VSTS feeds)
+                if (propertyName == "nuspec")
+                {
+                    XmlDocument xmlDoc = new XmlDocument();
+                    string xmlString = (string)psProperty.Value;
+
+                    if (string.IsNullOrWhiteSpace(xmlString))
+                    {
+                        throw new ArgumentNullException("xmlString");
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: Do we really want to throw exception here if the psProperty does not have value? If the repo server does not send us nuspec info in the body, should we log warning with meaningful user faced message instead? Also not all repo will has the problem you are fixing right? In that case we should not throw, yes?

File: Utility/DynamicJsonParser.cs (hunk 4 above)
Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: What's happening is that any html tag that specifies an encoding adds a non-printing character before the xml tag.  I'm not quite sure why this happens and it happens no matter what the actual encoding is.  I was able to reproduce this by manually copying and pasting the xml, so it's not related to the code itself.

File: Utility/DynamicJsonParser.cs [hunk 12]
```diff
@@ -135,6 +136,72 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                    if (xmlString[0] != '<')
+                    {
+                        var strIndex = xmlString.IndexOf('<');
+                        xmlString = xmlString.Substring(strIndex, xmlString.Length - 1);
+                    }
+
+                    try
+                    {
+                        XmlDocument xmlDoc = new XmlDocument();
+                        xmlDoc.LoadXml(xmlString);
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: xmlString can be null or empty here.

**Comment by alerickson** [comment]
@jianyunt please let me know if there are any more changes I should make.

File: Utility/DynamicJsonParser.cs [hunk 13]
```diff
@@ -135,6 +136,72 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                                        object actualPropertyVal = ConvertObject(propertyVal);
+                                        ((IDictionary<string, object>)actualObj)[property] = actualPropertyVal;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    catch (XmlException e)
+                    {
+                        throw new XmlException("XMl could not be loaded properly.", e);
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: @alerickson I don't think we need try-catch and rethrow here. I recommend we remove them. Later on this NuGetProvider should  catch it and log as warning/error. You may try it out to see if this is behavior. However if you want to use try-catch and rethrow, then add the `XMl could not be loaded properly.`  in the .resx file due to localization.
Your xmlDoc.LoadXml(xmlString) will fail if xmlString is null or empty.  What I meant in my previous comment is that, to be user friendly, we need to add an warning message saying something like "Cannot find valid nuspec content" .  Yes.  we need to add it to https://github.com/OneGet/NuGetProvider/blob/1f99213544a31226ed098f562af1e14ae5872a17/resources/Messages.resx
and throw this message for both line 164 and a new line before loadxml.  
It seems like overkill and this scenario never can happen. But the reality is these kind errors can happen because we do not control what server side can send us. Make sense?

File: resources/Microsoft.PackageManagement.NuGetProvider.Resources.Messages.resx [hunk 14]
```diff
@@ -0,0 +1,385 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (jianyunt) [inline resources/Microsoft.PackageManagement.NuGetProvider.Resources.Messages.resx]: @alerickson this file is auto-generated. Please do not check it in.

File: Utility/DynamicJsonParser.cs [hunk 15]
```diff
@@ -135,6 +137,73 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+
+                    if (string.IsNullOrWhiteSpace(xmlString))
+                    {
+                         var message = string.Format(Messages.InvalidNuspec, "xmlString");
+                         throw new InvalidDataException(message);
+                    }
+
+                    // Remove any characters that may come before the xml tag
+                    // Characters may appear before the start of an xml tag if xml encoding is specified
+                    if ((xmlString[0] != '<') && xmlString.Contains('<'))
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs:171]: @alerickson I don't think this logic works. You may consider something like following.
``` C#
if (string.Equals(propertyName, "nuspec", StringComparison.OrdinalIgnoreCase))
{
	
	string xmlString = psProperty.Value as string
	
	if (!string.IsNullOrWhiteSpace(xmlString))
	{

 		if ((xmlString[0] != '<') && xmlString.Contains('<'))
                {
                     var strIndex = xmlString.IndexOf('<');
                     xmlString = xmlString.Substring(strIndex, xmlString.Length - strIndex);
               }

       }

      	 if (string.IsNullOrWhiteSpace(xmlString))
         {
                var message = string.Format(Messages.InvalidNuspec, "xmlString");
                throw new InvalidDataException(message);
         }


....


}
```

File: Utility/DynamicJsonParser.cs [hunk 16]
```diff
@@ -135,6 +137,73 @@ private static dynamic GetDynamic(PSObject obj, string numericalPrefix = "n", bo
...
+                    {
+                        var strIndex = xmlString.IndexOf('<');
+                        xmlString = xmlString.Substring(strIndex, xmlString.Length - 1);
```
Comment (jianyunt) [inline Utility/DynamicJsonParser.cs]: always 1?

File: Utility/DynamicJsonParser.cs (hunk 15 above)
Comment (alerickson) [reply Utility/DynamicJsonParser.cs:171]: I did change it from what we initially discussed because I realized a string such as "????" 
would throw an index out of bounds error which checking for the index of '<'.  That means there shouldn't be a situation in which a string would hit the second "if (string.IsNullOrWhiteSpace(xmlString))" condition.

File: Utility/DynamicJsonParser.cs (hunk 16 above)
Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: I'm assuming the nuspec content is valid from '<' on since what's contained the in the nuspec file can vary.

File: resources/Microsoft.PackageManagement.NuGetProvider.Resources.Messages.resx (hunk 14 above)
Comment (alerickson) [reply resources/Microsoft.PackageManagement.NuGetProvider.Resources.Messages.resx]: I'll remove it

File: Utility/DynamicJsonParser.cs (hunk 16 above)
Comment (jianyunt) [reply Utility/DynamicJsonParser.cs]: That's possibly the reason why you are getting indexoutofbound for string like "???".  Should we  `xmlString = xmlString.Substring(strIndex, xmlString.Length - strIndex` instead?

Comment (alerickson) [reply Utility/DynamicJsonParser.cs]: The index out of bounds error only happens for cases where '<' doesn't exist in the string.  I have substring(strIndex, xmlString.Length - 1)  so that if I have a situation such as  "?<expected nuspec content>"  (which is what I was getting when testing with Kusto Tools because of a weird thing that happens when xml encoding is specified in the xml tag) I can just remove the unwanted '?' that appears before the expected nuspec content.

**Review by jianyunt — APPROVED** [review]
